src/error.rs

```diff
@@ -0,0 +1,130 @@
+
+/// An error that occurs during certificate validation or name validation.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Error {
+    /// The encoding of some ASN.1 DER-encoded item is invalid.
+    BadDER,
+
+    /// The encoding of an ASN.1 DER-encoded time is invalid.
+    BadDERTime,
+
+    /// A CA certificate is veing used as an end-entity certificate.
+    CAUsedAsEndEntity,
+
+    /// The certificate is expired; i.e. the time it is being validated for is
+    /// later than the certificate's notAfter time.
+    CertExpired,
+
+    /// The certificate is not valid for the name it is being validated for.
+    CertNotValidForName,
+
+    /// The certificate is not valid yet; i.e. the time it is being validated
+    /// for is earlier than the certificate's notBefore time.
+    CertNotValidYet,
+
+    /// An end-entity certificate is being used as a CA certificate.
+    EndEntityUsedAsCA,
+
+    /// An X.509 extension is invalid.
+    ExtensionValueInvalid,
+
+    /// The certificate validity period (notBefore, notAfter) is invalid; e.g.
+    /// the notAfter time is earlier than the notBefore time.
+    InvalidCertValidity,
+
+    /// The signature is invalid for the given public key.
+    InvalidSignatureForPublicKey,
+
+    /// The certificate violates one or more name constraints.
+    NameConstraintViolation,
+
+    /// The certificate violates one or more path length constraints.
+    PathLenConstraintViolated,
+
+    /// The algorithm in the TBSCertificate "signature" field of a certificate
+    /// does not match the algorithm in the signature of the certificate.
+    SignatureAlgorithmMismatch,
+
+    /// The certificate is not valid for the Extended Key Usage for which it is
+    /// being validated.
+    RequiredEKUNotFound,
+
+    /// A valid issuer for the certificate could not be found.
+    UnknownIssuer,
+
+    /// The certificate is not a v3 X.509 certificate.
+    UnsupportedCertVersion,
+
+    /// The certificate contains an unsupported critical extension.
+    UnsupportedCriticalExtension,
+
+    /// The signature's algorithm does not match the algorithm of the public
+    /// key it is being validated for. This may be because the public key
+    /// algorithm's OID isn't recognized (e.g. DSA), or the public key
+    /// algorithm's parameters don't match the supported parameters for that
+    /// algorithm (e.g. ECC keys for unsupported curves), or the public key
+    /// algorithm and the signature algorithm simply don't match (e.g.
+    /// verifying an RSA signature with an ECC public key).
+    UnsupportedSignatureAlgorithmForPublicKey,
+
+    /// The signature algorithm for a signature is not in the set of supported
+    /// signature algorithms given.
+    UnsupportedSignatureAlgorithm,
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		use self::Error::*;
+
+		f.write_str(match self {
+			BadDER =>
+				"The encoding of some ASN.1 DER-encoded item is invalid",
```

Comment:
I intentionally never include human-oriented strings like this in my projects. To see why, consider how useless this is for somebody who doesn't read English.

Instead, let's just implement `core::fmt::Display` in terms of `core::fmt::Debug` so that the output can be used for a properly internationalized (by the application) output.